### PR TITLE
KAFKA-13785: [2/N][emit final] add processor metadata to be committed with offset

### PR DESCRIPTION
**Pull request by lihaosky**
Add processor metadata to be committed with offset to broker. Adding this so that we can store last processed window time gracefully.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: formatting/indention

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Not sure if I can follow? Why is this a key-value pair? And why are the types `<Bytes, byte[]>`?

The `OffsetAndMetadata` interface takes a `String metadata` argument. That is why we encode the current `streamTime` we store as metadata as `String` encoded using base64.

If we assume that there might be multiple processors inside a task, I understand that we need a mapping from processor name to metadata, but the map should be `<String,String>` (or maybe `<String,Long>`)?

And we would encode it as concatenation of `<streamTime><numberOfEntries><key,value><key,value>` with key being the processor name, and value being the encoded timestamp (if we use `Long` in the map, we convert the long to base64 internally?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Why do we pass in ` TopicPartition`? Tasks are per partition already?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: What is the purpose of `merge` ? It seems it would be sufficient for a processor to just set it's own metadata?

The runtime can actually add the name of the processor internally?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Why do we need this?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Might be better to add a proper POJO maybe `StreamsMetadata` or something that wraps the `streamTime` Long plus `ProcessorMetadata` instead of using `KeyValue` ? We might add new fields later on what is easier to do for a new POJO.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Seems this will be removed as it's replace by `decodeTimestampAndMeta` that can handle old and new format?

**Review by mjsax — COMMENTED** [review]
I am wondering if we should make this public API (and add to the KIP?)

Otherwise, we need to case to an internal interface in the implementation? If we think this cast is ok, and we don't want to expand the scope of the KIP, I am also fine with it. Just wondering... We could also expose it as public API later on, but I think we should design it in a way that will allow us to expose it as public API later without major changes?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: > Why is this a key-value pair

My original intention is to make it flexible so that even different processor can commit/store multiple key/value pairs. So the key/value could be anything the processor choose. Now I think maybe we could also add a namespace to prevent collision of key from different processor. Namespace can be processor name.

> And why are the types <Bytes, byte[]>

Why not `<String, Long>`? I think `Long` is not flexible enough. For emit final use case, `Long` is fine. 
Why not `<String, String>`? I think it's more steps for processor to serialize it to `String` in some cases. e.g. For `Long`, processor needs to serialize it to `byte[]` and then `String`? So why not pass `byte[]` directly and serialize them finally to `String` for `OffsetAndMetadata`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: It's used in `StreamTask` line 1107 for serializing this object to `byte[]` before put into `OffsetAndMetadata`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Yeah. `TopicPartition` can be dropped.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: For `globalMetadata`, I'm thinking this should be committed to all `TopicPartition`. We need to merge the metadata from all `TopicPartition` to get correct `globalMetadata` in case some commits fail?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Right.

**Review by lihaosky — COMMENTED** [review]
> I am wondering if we should make this public API (and add to the KIP?)
Otherwise, we need to case to an internal interface in the implementation? If we think this cast is ok, and we don't want to expand the scope of the KIP, I am also fine with it. Just wondering... We could also expose it as public API later on, but I think we should design it in a way that will allow us to expose it as public API later without major changes?

I'm thinking of creating internal interface to reduce the scope of the KIP. Yeah, we should design it in a way for easy public API support

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Might be pre-mature optimization to make it generic. I would stick to just what we need. That is also why I think `<String,Long>` might just be good enough as we intend to only use it for "emit final".

> So why not pass byte[] directly and serialize them finally to String for OffsetAndMetadata?

My point is, that we should avoid to change the format/serialization twice, but only once. Either we accept `long` and let the runtime do `long -> String` directly. Or we just use `String` and the runtime has nothing to do but to concatenate string, and the processor needs to do the `long -> String` conversion. In both cases it's a single translation step. Introducing `byte[]` add a second translation step: what would be the advantage of having two steps instead of one?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: > this should be committed to all TopicPartition

Sound like we would commit it redundantly? Why store the same metadata for multiple partitions?

> We need to merge the metadata from all TopicPartition to get correct globalMetadata in case some commits fail?

Not sure if I can follow. If the commit fails, we would fall back to the previous offset including the previous metadata, right?

**Comment by lihaosky** [comment]
I plan to make changes. Don't review

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java]: nit: in AK it's common to omit `get` prefix in getters. The name should be `processorMetadataForKey`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:271]: Why do we need both "per key" and "full" getters/setters -- could we limit it to only one of both?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Why do we need this factory method? Isn't a public constructor not simpler?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Same here: why not just use a public constructor that take a `Map`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: nit: `metaDataBytes` (or similar to indicate it's not a `Metadata` object?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Seems we are calling `getBytes` and only take the length and throw away the serialized bytes and call `getBytes` below a second time.

Should we use `ByteArrayOutputStream` instead of `ByteBuffer` to avoid this double serialization (and also double looping through the map)?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: nit: `addMetadata` -> `put`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: nit `getMetadata` -> get`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: `needCommit` -> `needsCommit`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: `setNeedsCommit` -> `{@link #setNeedsCommit}`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java:131]: I assume it's on purpose to omit `needsCommit` flag? Might be worth to add a comment.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: It seems we need to guard this write and make a case decision during rolling upgrade -- it might only be save to use the new format after all instances are upgrade? Or do we consider using the new "emit final" feature a "breaking change" anyway and existing apps can only use it after a reset anyway?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:545]: Is setting the commit flag enough? Or would we need to commit the current metadata into the new partition(s) before starting to process any data? Atm, it seems we would only commit to the new partitions on the first commit after data was processed. Wondering if this is a race condition we need to close?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: It seems to be easier to just do:
```
final ProcessorMetadata finalProcessMetadata = new ProcessorMetadata()`
```

And always call `update()` blindly inside the loop. We also don't need the last check:
```
finalProcessMetadata = finalProcessMetadata == null ? ProcessorMetadata.emptyMetadata() : finalProcessMetadata;
```
as we know that it cannot be `null`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java]: Why using factory methods and not just public constructors?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java:59]: Still wondering, why we encode `processorMetadata` first as `byte[]` and convert it to `String` using Base64 here? Why not convert the `Map` to `String` directly`? The key is already String, so we would only use Base64 to convert the Long-values to String and concatenate all kv-pairs in some format (eg, CSV?)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorMetadataTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorMetadataTest.java:163]: Should we add a test that the commit flag has no impact on `equals` and `hashCode` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java:271]: This is used in restoration to set processor metadata as a whole after deserialization and merging. The per key setter is used by processor. Maybe we can get rid of `getProcessorMetadata` if it's not used.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: `-1L,` -> `RecordQueue.UNKNOWN` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: I thought it's clear to create emptyMetadata like this, but I'm ok to use public constructor as well.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: I can change to public constructor.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: The original purpose of this test seems to be testing "partial commits" (ie, for a sub-set of partitions)? Adding metadata make the partial commit a "full commit" and thus seems to change the test?

Als also not sure what the difference of this test compare to the newly added test below is?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: nit: `-1` -> `UNKNOWN` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java]: Good point!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadataTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadataTest.java]: I actually believe that the consumer contract is to ensure that the metadata `String` is never `null`? I think it's better to not have this test to not mask a potential bug?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java]: `encryptedString` cannot be `null` IIRC -- seems we should not check for this condition to no mask a potential consumer bug, but rather fail-fast.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Existing windowed aggregation not using "emit final" can't upgrade to it since store format is different. Yes, it's a breaking changing if user wants to use it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:545]: If new partitions are added, I think it's ok to not commit immediately since old partitions still have original metadata and they can be fetched/merged.

If there's partition deletion and system crashed before writing to new partitions, metadata will be gone. But I'm not sure if we need to handle that since user deleted old partitions containing metadata.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Good point. We can do that.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java]: I thought it's simpler to use factory for this simple pojo. But I can use public constructor if you think it's better.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java:59]: The reason is after converting single partitionTime to String, the length is not deterministic and we need other meta to encode it. Also I don't want to use Base64 multiple times.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:545]: I agree it might be a rare case. If we think it's not worth it and the risk to lose the metadata is tiny (and it's complicated to force a commit right away) I am fine with not closing this race condition.

**Review by mjsax — APPROVED** [review]
LGTM. We just need a Jira for KIP-825 before we can merge this PR.

**Comment by lihaosky** [comment]
Thanks @mjsax ! Created Jira: https://issues.apache.org/jira/browse/KAFKA-13785

**Pull request merged**
